tools/device-farm-runner/run_on_aws_devicefarm.py

```diff
@@ -0,0 +1,336 @@
+
+logging.basicConfig(level=logging.INFO)
+
+
+def parse_args() -> Any:
+    from argparse import ArgumentParser
+
+    parser = ArgumentParser("Run iOS tests on AWS Device Farm")
+    parser.add_argument(
+        "--project-arn", type=str, required=True, help="the ARN of the project on AWS"
+    )
+    parser.add_argument(
+        "--app-file", type=str, required=True, help="the iOS ipa app archive"
+    )
+    parser.add_argument(
+        "--xctest-file",
+        type=str,
+        required=True,
+        help="the XCTest suite to run",
+    )
+    parser.add_argument(
+        "--name-prefix",
+        type=str,
+        required=True,
+        help="the name prefix of this test run",
+    )
+    parser.add_argument(
+        "--device-pool-arn",
+        type=str,
+        default=DEFAULT_DEVICE_POOL_ARN,
+        help="the name of the device pool to test on",
+    )
+    parser.add_argument(
+        "--workflow-id",
+        type=str,
+        default="invalid-workflow-id",
+        help="the workflow run ID",
+    )
+    parser.add_argument(
+        "--workflow-attempt",
+        type=int,
+        default=0,
+        help="the workflow run attempt",
+    )
+
+    return parser.parse_args()
+
+
+def upload_file(
+    client: Any,
+    project_arn: str,
+    prefix: str,
+    filename: str,
+    filetype: str,
+    mime: str = "application/octet-stream",
+) -> str:
+    """
+    Upload the app file and XCTest suite to AWS
+    """
+    r = client.create_upload(
+        projectArn=project_arn,
+        name=f"{prefix}_{os.path.basename(filename)}",
+        type=filetype,
+        contentType=mime,
+    )
+    upload_name = r["upload"]["name"]
+    upload_arn: str = r["upload"]["arn"]
+    upload_url = r["upload"]["url"]
+
+    with open(filename, "rb") as file_stream:
+        info(f"Uploading {filename} to Device Farm as {upload_name}...")
+        r = requests.put(upload_url, data=file_stream, headers={"content-type": mime})
+        if not r.ok:
+            raise Exception(f"Couldn't upload {filename}: {r.reason}")
+
+    start_time = datetime.datetime.now()
+    # Polling AWS till the uploaded file is ready
+    while True:
+        waiting_time = datetime.datetime.now() - start_time
+        if waiting_time > datetime.timedelta(seconds=MAX_UPLOAD_WAIT_IN_SECOND):
+            raise Exception(
+                f"Uploading {filename} is taking longer than {MAX_UPLOAD_WAIT_IN_SECOND} seconds, terminating..."
+            )
+
+        r = client.get_upload(arn=upload_arn)
+        status = r["upload"].get("status", "")
+
+        info(f"{filename} is in state {status} after {waiting_time}")
+
+        if status == "FAILED":
+            raise Exception(f"Couldn't upload {filename}: {r}")
+        if status == "SUCCEEDED":
+            break
+
+        time.sleep(POLLING_DELAY_IN_SECOND)
+
+    return upload_arn
+
+
+def is_success(result: Optional[str]) -> bool:
+    return bool(result and result == "PASSED")
```

Comment:
Nit (None is not equal to any string)
```suggestion
    return result == "PASSED"
```